optaplanner-examples/src/main/java/org/optaplanner/examples/pas/solver/score/PatientAdmissionMoveConstraintProvider.java

```diff
@@ -0,0 +1,180 @@
+                preferredMaximumRoomCapacityConstraint(constraintFactory),
+                departmentSpecialismConstraint(constraintFactory),
+                roomSpecialismNotExistsConstraint(constraintFactory),
+                roomSpecialismNotFirstPriorityConstraint(constraintFactory),
+                preferredPatientEquipmentConstraint(constraintFactory)
+        };
+    }
+
+    public Constraint sameBedInSameNightConstraint(ConstraintFactory constraintFactory) {
+        return constraintFactory.fromUniquePair(BedDesignation.class)
+                .filter((left, right) -> left.getBed() != null
+                        && right.getBed() != null
+                        && left.getBed() == right.getBed()
+                        && left.getAdmissionPart().calculateSameNightCount(right.getAdmissionPart()) > 0)
+                .penalize("sameBedInSameNight", HardMediumSoftScore.ofHard(1000),
+                        (leftBd, rightBd) -> leftBd.getAdmissionPart().calculateSameNightCount(rightBd.getAdmissionPart()));
+    }
+
+    public Constraint femaleInMaleRoomConstraint(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(BedDesignation.class)
+                .filter(bd -> bd.getPatientGender() == Gender.FEMALE
+                        && bd.getRoomGenderLimitation() == GenderLimitation.MALE_ONLY)
+                .penalize("femaleInMaleRoom", HardMediumSoftScore.ofHard(50), BedDesignation::getAdmissionPartNightCount);
+    }
+
+    public Constraint maleInFemaleRoomConstraint(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(BedDesignation.class)
+                .filter(bd -> bd.getPatientGender() == Gender.MALE
+                        && bd.getRoomGenderLimitation() == GenderLimitation.FEMALE_ONLY)
+                .penalize("maleInFemaleRoom", HardMediumSoftScore.ofHard(50), BedDesignation::getAdmissionPartNightCount);
+    }
+
+    public Constraint differentGenderInSameGenderRoomInSameNightConstraint(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(BedDesignation.class)
+                .filter(bd -> bd.getRoomGenderLimitation() == GenderLimitation.SAME_GENDER
+                        && bd.getBed() != null)
+                .join(BedDesignation.class,
+                        equal(BedDesignation::getRoom),
+                        lessThan(BedDesignation::getId),
+                        filtering((left, right) -> right.getRoomGenderLimitation() == GenderLimitation.SAME_GENDER
+                                && left.getAdmissionPart().calculateSameNightCount(right.getAdmissionPart()) > 0
+                                && left.getPatient().getGender() != right.getPatient().getGender()))
+                .penalize("differentGenderInSameGenderRoomInSameNight", HardMediumSoftScore.ofHard(1000),
+                        (left, right) -> left.getAdmissionPart().calculateSameNightCount(right.getAdmissionPart()));
```

Comment:
Considering how often `calculateSameNightCount()` is used, I'd consider caching the calculation somehow. But maybe it's not a problem, doing a quick profiler run would tell you.

Reply:
Is that still relevant?

Reply:
No.